docs: cruise control monitoring (#5905)
strimzi/strimzi-kafka-operator: documentation/modules/metrics/con-metrics-cruise-control.adoc

// Module included in the following assemblies:
//
// metrics/assembly-metrics.adoc

[id='con-metrics-cruise-control-{context}']
= Monitoring Cruise Control operations

[role="_abstract"]
Kafka metrics are gathered by Cruise Control in order to track the utilization of brokers, topics and partitions.
Cruise Control also provides a set of metrics for monitoring its own performance.

> The Cruise Control metrics reporter collects raw metrics data from Kafka brokers, which is produced to auto-created Cruise Control topics.

Reviewer: This suggests Kafka auto-creation does it which is not true. We should probably say something more like _which are automatically created by Cruise Control_

The metrics are used to link:{BookURLUsing}#proc-generating-optimization-proposals-str[generate optimization proposals for Kafka clusters^].

Cruise Control metrics are available for real-time monitoring of Cruise Control operations.

> For example, you can use metrics to monitor the status of rebalancing operations that are running or provide alerts on any anomalies that are detected in its performance.

Reviewer: Are we now talking about the Strimzi Prometheus metrics? Is this included in the examples and Grafana dashboards? Or where is it available?

Reviewer 2: This refers to the Cruise Control metrics scraped by Prometheus [1] We have a Grafana  dashboard here [2]

[1] https://github.com/linkedin/cruise-control/wiki/Sensors
[2] https://github.com/strimzi/strimzi-kafka-operator/blob/main/examples/metrics/grafana-dashboards/strimzi-cruise-control.json

Author: Changed to _For example, you can use Cruise Control_ metrics

For a full list of available Cruise Control metrics, which are known as _sensors_, see the {CruiseControlSensorsDocs}.

== Exposing Cruise Control metrics

If you want to expose metrics on Cruise Control operations, configure the `Kafka` resource xref:proc-metrics-kafka-deploy-options-{context}[to deploy Cruise Control metrics configuration].
You can use your own configuration or an example file provided by Strimzi called `kafka-cruise-control-metrics.yaml`.

You add the configuration to the `metricsConfig` of the `CruiseControl` property in the `Kafka` resource.
The configuration enables JMX for Cruise Control and uses the {JMXExporter} to expose Cruise Control metrics through an HTTP endpoint.
The HTTP endpoint is scraped by the Prometheus server.

> .Example metrics configuration for Cruise Control

Author: @kyguy Not sure we need to show this example, as we link to the deploying prometheus metrics config procedure. Happy to keep for context

Reviewer 2: No harm in keeping it, if you are happy to keep it I am too!

  metadata:
    name: my-cluster
  Spec:
    # ...
    cruiseControl:
      # ...
      metricsConfig:
        type: jmxPrometheusExporter
        valueFrom:
          configMapKeyRef:
            name: cruise-control-metrics
            key: metrics-config.yml
  ---
  kind: ConfigMap
  apiVersion: v1
  metadata:
    name: cruise-control-metrics
    labels:
      app: strimzi
  data:
    metrics-config.yml: |
    # _metrics configuration..._
----

== Viewing Cruise Control metrics

Having exposed Cruise Control metrics, you can use a monitoring tool to view the information on the data.
Strimzi provides an xref:assembly-metrics-config-files-str[example Grafana dashboard] to provide visualizations of Cruise Control metrics.
The dashboard is a JSON file called `strimzi-cruise-control.json`.
The exposed metrics provide the monitoring data when you xref:proc-metrics-grafana-dashboard-str[enable the Grafana dashboard].

=== Monitoring Balancedness scores

Cruise Control metrics include balancedness scores.
Balancedness is the measure of how evenly a workload is distributed in a Kafka cluster.

The Cruise Control metric for balancedness score (`balancedness-score`) differs from the

> the balancedness score in the `KafkaRebalance` resource. The former is calculated using `anomaly.detection.goals`, while the latter is calculated using `default.goals`.

Reviewer: This might deserve some more explanation what `anomaly.detection.goals` and `default.goals` are?

Reviewer 2: @PaulRMellor The values of these balancedness scores  are calculated using different goal lists specified in the spec.cruiseControl.config` of the Kafka custom resource. The former is calculated using `anomaly.detection.goals`, while the latter is calculated using `default.goals`.

Author: @kyguy Changed to:

> The Cruise Control metric for balancedness score (`balancedness-score`) differs from the balancedness score in the `KafkaRebalance` resource. Cruise Control calculates each score using different goals.
> The former uses `anomaly.detection.goals`, while the latter uses `default.goals`.
> The goals are specified in the `spec.cruiseControl.config` of the `Kafka` custom resource.

> NOTE: Refreshing a `KafkaRebalance` resource fetches the latest state of a Kafka cluster. The refresh also requests an optimization proposal from Cruise Control. Continual refreshes can impact performance monitoring in certain circumstances. For example, the KafkaRebalance `goals` might differ from the `default.goals` used by Cruise Control.

Reviewer: Are these `default.goals` from the Kafka CR? Or is this something else?

Reviewer 2: Kafka `default.goals` (the `default.goals` specified in the `spec.cruiseControl.config` of the Kafka resource )

Author: @kyguy  I've changed to _For example, the `goals` configured in `KafkaRebalance` resource might differ from the `default.goals` used by Cruise Control._  I think this is what we're trying to say? Or should the first `goals` be `default.goals` too?  Can we say how the refreshes impact this? I'm not entirely sure we're clear on what's happening here.

Reviewer 2: Yes so Cruise Control calculates and caches optimization proposals based on the goals configured in the `default.goals` of the Kafka resource ahead of time. This cached optimization proposal will be fetched for a `KafkaRebalance` resource if :
1. The goals configured in the `goals` section of the KafkaRebalance resource match the goals configured in the `default.goals` section of the Kafka resource.
2.  No `goals`  section is specified  in the KafkaRebalance resource

Otherwise, the `KafkaRebalance` resource cannot use the cached optimzation proposal from Cruise Control and Cruise Control must generated a new optimzation  proposal on the fly based on the the goals configured in the `goals` section of the KafkaRebalance. This is where the extra performance cost comes from.

Author: Thanks @kyguy 
Here's the latest. Trying to keep it simple.

![image](https://user-images.githubusercontent.com/47596553/143904565-ed68b914-b489-4ba0-9ab6-c7825a307d82.png)

=== Alerts on anomaly detection

Cruise Control’s anomaly detection is enabled for use with AMQ Streams.
You can use anomaly detection when monitoring Cruise Control metrics.
Cruise Control’s anomaly notifier provides alerts on broker failures and other conditions that block the generation of optimization goals, as well as any actions taken.

> To route Cruise Control alerts to a notification service, use an alerting tool like Prometheus Alertmanager.

Author: Assuming we can set up alertmanager or similar for anomaly alerts

Reviewer 2: For sure it's possible but we don't have any examples for it in the Strimzi repo

Author: Okay. Can we keep this as is?

Reviewer 2: I think that would be fine

> To route Cruise Control alerts to a notification service, use an alerting tool like Prometheus Alertmanager.

Reviewer: Do we have any alert examples in the alert samples which we have?

Reviewer 2: We do not

Reviewer: This sounds weird. Who generates the alerts? It seems we suggest that Cruise Control generates the alerts. But how do they then get into the Prometheus AlertManager to alert about them? Or is the user expected to generate the alerts based on the Prometheus metrics from Cruise Control? That makes it clear how do they get into Prometheus Alertmanager. But we should not say that Cruise Control generates the alerts. I think this needs to be more clear which compoenent does what and how to use them.

Reviewer 2: > Or is the user expected to generate the alerts based on the Prometheus metrics from Cruise Control

This is the way

Reviewer 3: Should we point somewhere or providing more information about what are these metrics that can be scraped by Prometheus for defining alerts?

Reviewer 3: Talking about alerts in CC I found this in the Wiki : https://github.com/linkedin/cruise-control/wiki/Configure-notifications
It seems that CC has some built-in mechanisms to sends alerts to Slack, MS Teams, ...
How does it cope with what we are saying about Prometheus metrics exported by CC and then using Prometheus alert manager? I cannot find any info about it.
The only support about Prometheus that I see in CC is about a "Prometheus sampler", so instead of having the metrics sampler running on the broker and getting the JMX metrics, there is a sampler that get metrics from Prometheus endpoint, if your broker as per Strimzi, expose such a metrics in Prometheus format.
@kyguy can you point me to some reference about Prometheus Alertmanager with CC?

Reviewer 2: >  Should we point somewhere or providing more information about what are these metrics that can be scraped by Prometheus for defining alerts?

In the summary section of this PR we point to the Cruise Control "Sensors" Wiki [1] which provides Prometheus scrapable metrics names/descriptions [1] We should add the link here or instead for better context.

> How does it cope with what we are saying about Prometheus metrics exported by CC and then using Prometheus alert manager? I cannot find any info about it.

TBH I have not read about or used the Prometheus AlertManager not even in the context of Cruise Control. Since it was included here I assumed/misunderstood that Prometheus AlertManager could be used to scrape the Prometheus endpoint and define rules based on those metrics. But I think I was getting confused with the "Prometheus sampler" you mentioned. We should either swap out the mention of the Prometheus AlertManager with a tool that can create alerts using the scraped Promethues metrics or remove mention all together!

For the record, Cruise Control’s anomaly notifier provides _data_ not alerts. The _data_ could be used to create alerts with some third-party tool

[1] https://github.com/linkedin/cruise-control/wiki/Sensors

Reviewer 3: I think we should make clear what already is in the link you pasted above ...

> You can easily expose Cruise Control metrics by enabling Java Management Extensions (JMX), and collect them by using your favourite JMX collection tool (e.g - jmx_exporter).

So that CC metrics have to be enabled but they are in JMX format and then you can expose them in Prometheus using a tool like jmx_exporter. After that you can put Prometheus in place and Prometheus Alertmanager to generate alerts.
Anyway, yes CC exposes/provides data/metrics not alerts.

Author: Thanks @ppatierno  and @kyguy 

Intro:

> You can expose Cruise Control metrics by enabling JMX for Cruise Control.
> For a full list of available Cruise Control metrics, which are known as _sensors_, see the {CruiseControlSensorsDocs}.

We mention this again in _Exposing Cruise Control metrics_:

> The configuration enables JMX for Cruise Control and uses the {JMXExporter} to expose Cruise Control metrics through an HTTP endpoint.

Alerts on anomaly detection:

> Cruise Control’s anomaly notifier provides metrics data for conditions that block the generation of optimization goals, such as broker failures, as well as any actions taken.
> If you want more visibility, you can use these metrics to set up alerts that send out notifications.
> If you are using Prometheus, you can use Prometheus Alertmanager to generate alerts.

Reviewer 2: >  If you are using Prometheus, you can use Prometheus Alertmanager to generate alerts.

From what I understand, `Prometheus AlertManager` [1] _manages_ existing alerts, it does generate alerts. I think Paolo meant to suggest "put Prometheus in place _of_ Prometheus Alertmanager " so saying something like "You can use Prometheus to generate alerts".

> The Alertmanager handles alerts sent by client applications such as the Prometheus server.

[1] https://prometheus.io/docs/alerting/latest/alertmanager/

Reviewer 3: @kyguy yes. Prometheus provides a way to define alerts. Alertmanager is a tool to use them and forward to different platform like Slack, emails and so on.

Author: Okay. I've changed the Prometheus bit to:

> If you want more visibility, you can use these metrics to set up alerts and send out notifications.
> You can use Prometheus to generate alerts.
> Prometheus Alertmanager can then route the alerts through a notification channel.

Reviewer 2: Sorry Paul, I was confusing the terminology and mistakenly told you:

> For the record, Cruise Control’s anomaly notifier provides data not alerts.

The anomaly notifier _does_ provide alerts. You were right about this and I was wrong.

The Anomaly Detector and Anomaly Notifier [1] are two different things

- **Anomaly Detector:** provides data based on things like broker failures, and other other conditions.
- **Anomaly Notifier**: provides alerts (You were correct about this Paul! Sorry for asking you to change it). As Paolo pointed out before, CC has built in support for sending alerts to Slack,  MSTeams, Alerta [2]. 

However, the Anomaly Notifier does not support Prometheus. But as you wrote, you can set up Prometheus to scrape and create alerts from the data provided by the Anomaly Detector and set up Prometheus Alert Manager to manage alerts.

[1] https://github.com/linkedin/cruise-control/wiki/Pluggable-Components#anomaly-notifier
[2] https://github.com/linkedin/cruise-control/wiki/Configure-notifications

You can use your own configuration or an example file provided by Strimzi called `kafka-cruise-control-metrics.yaml`.

You add the configuration to the `metricsConfig` of the `CruiseControl` property in the `Kafka` resource.
The configuration enables JMX for Cruise Control and uses the {JMXExporter} to expose Cruise Control metrics through an HTTP endpoint.
The HTTP endpoint is scraped by the Prometheus server.

.Example metrics configuration for Cruise Control
[source,yaml,subs="+quotes,attributes"]
----
  apiVersion: {KafkaApiVersion}
  kind: Kafka
  metadata:
    name: my-cluster
  Spec:
    # ...
    cruiseControl:
      # ...
      metricsConfig:
        type: jmxPrometheusExporter
        valueFrom:
          configMapKeyRef:
            name: cruise-control-metrics
            key: metrics-config.yml
  ---
  kind: ConfigMap
  apiVersion: v1
  metadata:
    name: cruise-control-metrics
    labels:
      app: strimzi
  data:
    metrics-config.yml: |
    # _metrics configuration..._
----

== Viewing Cruise Control metrics

> Having exposed Cruise Control metrics, you can use a monitoring tool to view the information on the data.

Reviewer: I think _a monitoring tool_ suggest some specific monitoring tools exist for this. I would just say something like _you can use your Prometheus or compatible monitoring system to use the metrics_.

Author: Changed to:

> Having exposed Cruise Control metrics, you can use Prometheus or another suitable monitoring system view information on the metrics data.

The dashboard is a JSON file called `strimzi-cruise-control.json`.
The exposed metrics provide the monitoring data when you xref:proc-metrics-grafana-dashboard-str[enable the Grafana dashboard].

=== Monitoring Balancedness scores

Cruise Control metrics include balancedness scores.
Balancedness is the measure of how evenly a workload is distributed in a Kafka cluster.

The Cruise Control metric for balancedness score (`balancedness-score`) differs from the balancedness score in the `KafkaRebalance` resource.
Cruise Control calculates each score using different goals.
The former uses `anomaly.detection.goals`, while the latter uses `default.goals`.
The goals are specified in the `spec.cruiseControl.config` of the `Kafka` custom resource.

NOTE: Refreshing a `KafkaRebalance` resource fetches the latest state of a Kafka cluster.
The refresh also requests an optimization proposal from Cruise Control. Continual refreshes can impact performance monitoring in certain circumstances.
For example, the `goals` configured in the `KafkaRebalance` resource might differ from the `default.goals` used by Cruise Control.

=== Alerts on anomaly detection

> Cruise Control’s anomaly detection is enabled for use with AMQ Streams.

Reviewer: ```suggestion
Cruise Control’s anomaly detection is enabled for use with Strimzi.
```

You can use anomaly detection when monitoring Cruise Control metrics.

> Cruise Control’s anomaly notifier provides alerts on broker failures and other conditions that block the generation of optimization goals, as well as any actions taken.

Reviewer 2: alerts -> data

Following Jakub's comment on what exactly creates/sends "alerts". Cruise Control anomaly notifier provides metrics/data for an external alertmanager to alerts on

Author: Changed to:

> Cruise Control’s anomaly notifier provides metrics data for conditions that block the generation of optimization goals, such as broker failures. An alerting tool like Prometheus Alertmanager can use these metrics to send notifications about the events.

// Module included in the following assemblies:
//
// metrics/assembly-metrics.adoc

[id='con-metrics-cruise-control-{context}']
= Monitoring Cruise Control operations

[role="_abstract"]
Kafka metrics are gathered by Cruise Control in order to track the utilization of brokers, topics and partitions.
Cruise Control also provides a set of metrics for monitoring its own performance.

> The Cruise Control metrics reporter collects raw metrics data from Kafka brokers, which is produced to topics that are automatically created by Cruise Control.

Author: Changed to:

> The Cruise Control metrics reporter collects raw metrics data from Kafka brokers.
> The data is produced to topics that are automatically created by Cruise Control.

The metrics are used to link:{BookURLUsing}#proc-generating-optimization-proposals-str[generate optimization proposals for Kafka clusters^].

Cruise Control metrics are available for real-time monitoring of Cruise Control operations.
For example, you can use Cruise Control metrics to monitor the status of rebalancing operations that are running or provide alerts on any anomalies that are detected in its performance.

You can expose Cruise Control metrics by enabling JMX for Cruise Control.
For a full list of available Cruise Control metrics, which are known as _sensors_, see the {CruiseControlSensorsDocs}.

== Exposing Cruise Control metrics

If you want to expose metrics on Cruise Control operations, configure the `Kafka` resource xref:proc-metrics-kafka-deploy-options-{context}[to deploy Cruise Control metrics configuration].

> You can use your own configuration or an example file provided by Strimzi called `kafka-cruise-control-metrics.yaml`.

Author: Changed to:

> You can use your own configuration or use the example `kafka-cruise-control-metrics.yaml` file provided by Strimzi.

Strimzi provides an xref:assembly-metrics-config-files-str[example Grafana dashboard] to provide visualizations of Cruise Control metrics.
The dashboard is a JSON file called `strimzi-cruise-control.json`.
The exposed metrics provide the monitoring data when you xref:proc-metrics-grafana-dashboard-str[enable the Grafana dashboard].

=== Monitoring Balancedness scores

Cruise Control metrics include a balancedness score.
Balancedness is the measure of how evenly a workload is distributed in a Kafka cluster.

The Cruise Control metric for balancedness score (`balancedness-score`) differs from the balancedness score in the `KafkaRebalance` resource.
Cruise Control calculates each score using different goals.
The former uses `anomaly.detection.goals`, while the latter uses `default.goals`.
The goals are specified in the `spec.cruiseControl.config` of the `Kafka` custom resource.

[NOTE]
====
Refreshing the `KafkaRebalance` resource fetches an optimization proposal.
The latest cached optimization proposal is fetched under one of the following conditions:

> * KafkaRebalance `goals` match the goals configured in the `default.goals`section of the `Kafka` resource

Author: We leave without in the doc set if not full sentences

// Module included in the following assemblies:
//
// metrics/assembly-metrics.adoc

[id='con-metrics-cruise-control-{context}']
= Monitoring Cruise Control operations

[role="_abstract"]
Kafka metrics are gathered by Cruise Control in order to track the utilization of brokers, topics, and partitions.
Cruise Control also provides a set of metrics for monitoring its own performance.

The Cruise Control metrics reporter collects raw metrics data from Kafka brokers.
The data is produced to topics that are automatically created by Cruise Control.
The metrics are used to link:{BookURLUsing}#proc-generating-optimization-proposals-str[generate optimization proposals for Kafka clusters^].

Cruise Control metrics are available for real-time monitoring of Cruise Control operations.
For example, you can use Cruise Control metrics to monitor the status of rebalancing operations that are running or provide alerts on any anomalies that are detected in an operation's performance.

> You can expose Cruise Control metrics by enabling JMX for Cruise Control.

Reviewer 2: Thanks Jakub, I misunderstood this but you are right, we don't expose JMX since we never set KAFKA_JMX_OPTS [1] for the Cruise Control container. We only enable the Prometheus JMX Exporter [2] which "exposes Cruise Control metrics through an HTTP endpoint"

[1] https://github.com/linkedin/cruise-control/blob/migrate_to_kafka_2_4/kafka-cruise-control-start.sh#L65
[2] https://github.com/strimzi/strimzi-kafka-operator/blob/main/docker-images/kafka/cruise-control-scripts/cruise_control_run.sh#L28

> You can expose Cruise Control metrics by enabling JMX for Cruise Control.

Author: I've expanded on this a bit more, as follows.
Links to the procedure _Deploying Prometheus metrics configuration_

> You expose Cruise Control metrics by enabling the {JMXExporter} in the Cruise Control configuration.
> You can xref:proc-metrics-kafka-deploy-options-str[deploy additional metrics configuration in a ConfigMap].
> ```
> .Enabling the Prometheus JMX Exporter for Cruise Control
> [source,yaml,subs="+quotes,attributes"]
> ----
> apiVersion: {KafkaApiVersion}
> kind: Kafka
> metadata:
>   name: my-cluster
> spec:
>   # ...
>   cruiseControl:
>     metricsConfig:
>       type: jmxPrometheusExporter
>       valueFrom:
>         configMapKeyRef:
>           name: cruise-control-metrics
>           key: metrics-config.yml
> ----
> 
> ```

For a full list of available Cruise Control metrics, which are known as _sensors_, see the {CruiseControlSensorsDocs}.

== Exposing Cruise Control metrics

> If you want to expose metrics on Cruise Control operations, configure the `Kafka` resource xref:proc-metrics-kafka-deploy-options-{context}[to deploy Cruise Control metrics configuration].

Author: Changed to _to deploy Cruise Control and enable Prometheus metrics in the deployment_

You can use your own configuration or use the example `kafka-cruise-control-metrics.yaml` file provided by Strimzi.

You add the configuration to the `metricsConfig` of the `CruiseControl` property in the `Kafka` resource.

> The configuration enables JMX for Cruise Control and uses the {JMXExporter} to expose Cruise Control metrics through an HTTP endpoint.

Reviewer 2: Thanks Jakub, I misunderstood this but you are right, we don't expose JMX since we never set `KAFKA_JMX_OPTS` [1] for the Cruise Control container. We only enable the Prometheus JMX Exporter [2] which "exposes Cruise Control metrics through an HTTP endpoint"

[1] https://github.com/linkedin/cruise-control/blob/migrate_to_kafka_2_4/kafka-cruise-control-start.sh#L65
[2] https://github.com/strimzi/strimzi-kafka-operator/blob/main/docker-images/kafka/cruise-control-scripts/cruise_control_run.sh#L28

> The configuration enables JMX for Cruise Control and uses the {JMXExporter} to expose Cruise Control metrics through an HTTP endpoint.

Author: Changed to: _The configuration enables the {JMXExporter} to expose Cruise Control metrics through an HTTP endpoint._

----
  apiVersion: {KafkaApiVersion}
  kind: Kafka
  metadata:
    name: my-cluster
  Spec:
    # ...
    cruiseControl:
      # ...
      metricsConfig:
        type: jmxPrometheusExporter
        valueFrom:
          configMapKeyRef:
            name: cruise-control-metrics
            key: metrics-config.yml
  ---
  kind: ConfigMap
  apiVersion: v1
  metadata:
    name: cruise-control-metrics
    labels:
      app: strimzi
  data:
    metrics-config.yml: |
    # _metrics configuration..._
----

== Viewing Cruise Control metrics

After you expose the Cruise Control metrics, you can use Prometheus or another suitable monitoring system to view information on the metrics data.
Strimzi provides an xref:assembly-metrics-config-files-str[example Grafana dashboard] to display visualizations of Cruise Control metrics.
The dashboard is a JSON file called `strimzi-cruise-control.json`.
The exposed metrics provide the monitoring data when you xref:proc-metrics-grafana-dashboard-str[enable the Grafana dashboard].

=== Monitoring balancedness scores

Cruise Control metrics include a balancedness score.
Balancedness is the measure of how evenly a workload is distributed in a Kafka cluster.

> The Cruise Control metric for balancedness score (`balancedness-score`) differs from the balancedness score in the `KafkaRebalance` resource.
> Cruise Control calculates each score using different goals.
> The former uses `anomaly.detection.goals`, while the latter uses `default.goals`.
> The goals are specified in the `spec.cruiseControl.config` of the `Kafka` custom resource.

Reviewer: It sounds like the goals might differ, but also might not, or?

```suggestion
The Cruise Control metric for balancedness score (`balancedness-score`) might differ from the balancedness score in the `KafkaRebalance` resource.
Cruise Control calculates each score using `anomaly.detection.goals` which might not be the same as the goals used in the `KafkaRebalance` resource.
The `anomaly.detection.goals` goals are specified in the `spec.cruiseControl.config` of the `Kafka` custom resource.
```

[NOTE]
====
Refreshing the `KafkaRebalance` resource fetches an optimization proposal.
The latest cached optimization proposal is fetched if one of the following conditions applies:

* KafkaRebalance `goals` match the goals configured in the `default.goals`section of the `Kafka` resource
* KafkaRebalance `goals` are not specified

Otherwise, Cruise Control generates a new optimization proposal based on KafkaRebalance `goals`. If new proposals are generated with each refresh, this can impact performance monitoring.
====

=== Alerts on anomaly detection

Cruise Control’s anomaly detection is enabled for use with Strimzi.

> You can use anomaly detection when monitoring Cruise Control metrics.

Reviewer: I don't think this makes sense. It suggests that the user should somehow apply some anomaly detection tool or technique on the CC metrics data. But that contradicts the other parts of the docs. IMHO it should be deleted, or maybe it should be something like _Anomaly detection results are part of the Cruise Control metrics._ But I think it would be good to provide more details of what will be the metric names etc.

Author: I think we can lose those first two sentences.
And add another link to see the list of anomaly metrics. (We also have one in the intro)
So we have the following.
Not sure what you mean by adding more details on metric names?

> Cruise Control’s anomaly notifier provides metrics data for conditions that block the generation of optimization goals, such as broker failures, and any actions taken.
> If you want more visibility, you can use these metrics to set up alerts and send out notifications.
> You can use Prometheus to generate alerts.
> Prometheus Alertmanager can then route the alerts through a notification channel.
> 
> The {CruiseControlSensorsDocs} provides information on `AnomalyDetector` metrics.

Reviewer 2: I agree with your suggestions Paul but I think you missed the edits in your example in the above! The first two sentences are still there

Author: Thanks, Kyle. I haven't pushed the changes up yet. Waiting on clarification of JMX, which we have now.

Author: Updated as follows:

> **Alerts on anomaly detection**
> 
> Cruise control's _anomaly detector_ provides metrics data for conditions that block the generation of optimization goals, such as broker failures.
> If you want more visibility, you can use the metrics provided by the anomaly detector to set up alerts and send out notifications.
> You can set up Cruise Control’s _anomaly notifier_ to route alerts based on these metrics through a specified notification channel.
> Alternatively, you can set up Prometheus to scrape the metrics data provided by the anomaly detector and generate alerts.
> Prometheus Alertmanager can then route the alerts generated by Prometheus.
> 
> The {CruiseControlConfigDocs} provides information on `AnomalyDetector` metrics and the anomaly notifier.
